specs/src/interoperability/ethereum-bridge/ethereum_events_attestation.md

```diff
@@ -18,81 +34,126 @@ pub enum EthereumEvent {
+/eth_msgs/$msg_hash/voting_started_epoch: Epoch # epoch when the tally started
 ```
 
-`\$msg_hash` is the SHA256 digest of the Borsh serialization of the relevant
-`EthereumEvent`.
+Where `$msg_hash` is the SHA256 digest of the Borsh serialization of
+some `EthereumEvent`.
 
-Changes to this `/eth_msgs` storage subspace are only ever made by
-nodes as part of the ledger code based on the aggregate of votes
-by validators for specific events. That is, changes to
-`/eth_msgs` happen
-in block `n` in a deterministic manner based on the votes included in the
-block proposal for block `n`. Depending on the underlying Tendermint
-version, these votes will either be included as vote extensions or as
-protocol transactions.
+Changes to this `/eth_msgs` storage subspace are only ever made by nodes
+as part of the ledger code based on the aggregate of votes by validators
+for specific events. That is, changes to `/eth_msgs` happen in block `n`
+in a deterministic manner based on the votes included in the block
+proposal for block `n`. Depending on the underlying Tendermint version,
+these votes will either be included as vote extensions or as protocol
+transactions.
 
-The `/eth_msgs` storage subspace will belong
-to the `EthBridge` validity predicate. It should disallow any changes to
-this storage from wasm transactions.
+The `/eth_msgs` storage subspace will belong to the `EthBridge` validity
+predicate. It should disallow any changes to this storage from wasm
+transactions.
 
 ### Including events into storage
 
-For every Namada block proposal, block proposer should include the votes for
-events from other validators into their proposal. If the underlying Tendermint
-version supports vote extensions, consensus invariants guarantee that a
-quorum of votes from the previous block height can be included. Otherwise,
-validators can only submit votes by broadcasting protocol transactions,
-which comes with less guarantees (i.e. no consensus finality).
-
-The vote of a validator should include the events of the Ethereum blocks they
-have seen via their full node such that:
-1. It's correctly formatted.
-2. It's reached the required number of confirmations on the Ethereum chain
-
-Each event that a validator is voting to include must be individually signed by
-them. If the validator is not voting to include any events, they must still
-provide a signed empty vector of events to indicate this.
-
-The votes will include be a Borsh-serialization of something like
-the following.
-```rust
+For every Namada block proposal, the proposer should include the votes
+for events from other validators into their proposal. If the underlying
+Tendermint version supports vote extensions, consensus invariants
+guarantee that a quorum of votes from the previous block height can be
+included. Otherwise, validators can only submit votes by broadcasting
+protocol transactions, which comes with less guarantees (i.e. no
+consensus finality).
```

Comment:
My understanding of protocol transactions is that they are final, but that a proposer can choose to include or not include votes and protocol transactions independently, while with ABCI++ a proposer must include either the vote and the vote extension or neither. Is that accurate, or not? If it's accurate, we should clarify a bit.

Reply:
so, to clarify, a vote is just an abstract concept - in these cases, a vote can be made through a vote extension, or through a protocol transaction. a vote extension piggy backs off of BFT consensus, while a protocol transaction must be recorded on-chain, and go through the mempool (hence the delay in the latter).

you could claim that a proposer can censor votes, by not accounting for them, but this is also the case with votes in the sense of BFT consensus (each protocol step can be considered a vote that advances the state of consensus).

regarding protocol transactions being final in the sense of BFT consensus, I don't think they are, unless you do some weird on-chain 3-step consensus.

Reply:
hmm, I think I might be using the terms slightly differently. Protocol transactions, if included by some proposer in a block, are broadcast to the other validators, who execute them as part of that block, updating the state root accordingly, yes? That's all I mean by "finality" - once a protocol transaction is included in a block accepted by consensus, the state changes are final just as for regular transactions.

the difference which I understood - which might be wrong - is that I thought Tendermint had validators sign over the vote extension data in the vote itself, such that the proposer would have to choose one of two options:
1. don't include the (BFT consensus) vote, and don't include the vote extension data
2. include the BFT consensus vote, and include the vote extension data

with protocol transactions, the proposer has two more options:
3. include the BFT consensus vote, but _don't_ include the "vote extension" (protocol transaction) data
4. don't include the BFT consensus vote, but _do_ include the "vote extension" (protocol transaction) data

am I misunderstanding something here?